src/oauth.ts

```diff
@@ -0,0 +1,505 @@
+import { ProxyOAuthServerProvider, ProxyOptions } from '@modelcontextprotocol/sdk/server/auth/providers/proxyProvider.js';
+import { config } from './config.js';
+import { mcpAuthRouter } from '@modelcontextprotocol/sdk/server/auth/router.js';
+import { OAuthClientInformationFull, OAuthTokens } from '@modelcontextprotocol/sdk/shared/auth.js';
+import { AuthorizationParams } from '@modelcontextprotocol/sdk/server/auth/provider.js';
+import { AuthInfo } from '@modelcontextprotocol/sdk/server/auth/types.js';
+import { OAuthRegisteredClientsStore } from '@modelcontextprotocol/sdk/server/auth/clients.js';
+import { Request, Response } from 'express';
+import { logger } from './logger.js';
+import { Database } from 'better-sqlite3';
+import argon2 from '@node-rs/argon2';
+import { randomBytes } from 'crypto';
+
+// Custom provider that handles dynamic registration and maps to GitLab OAuth
+class GitLabProxyProvider extends ProxyOAuthServerProvider {
+  // Static async factory method
+  static async New(options: any): Promise<GitLabProxyProvider> {
+    // we put this here so we dont initialize this unless we are using the oauth provider
+    const Database = (await import('better-sqlite3')).default;
+    const db = new Database(config.GITLAB_OAUTH2_DB_PATH);
```

Comment:
Persistency should really be decoupled from the OAuth implementation. This means:
1. Move all db logic to the a separate file
2. Create a clean interface that hides the persistency logic

We could then even have different implementations for this interface (in-memory, sqlite, redis?). Or we could consider using a ORM.

Reply:
the persistence is specific to the oauth implementation